File: csharp/ToolGood.Algorithm/Internals/Functions/Value/Function_ValueText.cs
using System;
using System.Text;
using ToolGood.Algorithm.Enums;

namespace ToolGood.Algorithm.Internals.Functions.Value
{
	internal sealed class Function_ValueText : Function_0
	{
		private readonly Operand _value;

		public Function_ValueText(Operand value)
		{
			_value = value;
		}
		 
		public override string Name => "Value";

		public override Operand Evaluate(AlgorithmEngine engine, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			return _value;
		}
		public override void ToString(StringBuilder stringBuilder, bool addBrackets)
		{
			stringBuilder.Append('"');
			// 单遍转义,避免链式 Replace 多次创建中间字符串
			foreach(var c in _value.TextValue) {
				switch(c) {
					case '\\': stringBuilder.Append("\\\\"); break;
					case '"': stringBuilder.Append("\\\""); break;
					case '\r': stringBuilder.Append("\\r"); break;
					case '\n': stringBuilder.Append("\\n"); break;
					case '\t': stringBuilder.Append("\\t"); break;
					case '\0': stringBuilder.Append("\\0"); break;
					case '\v': stringBuilder.Append("\\v"); break;
					case '\a': stringBuilder.Append("\\a"); break;
					case '\b': stringBuilder.Append("\\b"); break;
					case '\f': stringBuilder.Append("\\f"); break;
					default: stringBuilder.Append(c); break;
				}
			}
			stringBuilder.Append('"');
		}
		public override OperandType GetResultType()
		{
			return OperandType.TEXT;
		}
	}

}

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/value/Function_ValueText.java
package toolgood.algorithm.internals.functions.value;

import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.functions.Function_0;

public final class Function_ValueText extends Function_0 {
    private final Operand _value;

    public Function_ValueText(Operand value) {
        _value = value;
    }

    @Override
    public String Name() {
        return "Value";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        return _value;
    }

    @Override
    public void ToString(StringBuilder stringBuilder, boolean addBrackets) {
        stringBuilder.append('"');
        String stringValue = _value.TextValue();
        stringValue = stringValue.replace("\\", "\\\\");
        stringValue = stringValue.replace("\r", "\\r");
        stringValue = stringValue.replace("\n", "\\n");
        stringValue = stringValue.replace("\t", "\\t");
        stringValue = stringValue.replace("\0", "\\0");
        stringValue = stringValue.replace("\u000B", "\\v");
        stringValue = stringValue.replace("\u0007", "\\a");
        stringValue = stringValue.replace("\b", "\\b");
        stringValue = stringValue.replace("\f", "\\f");
        stringValue = stringValue.replace("\"", "\\\"");
        stringBuilder.append(stringValue);
        stringBuilder.append('"');
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }
}
